src/efftester.ml

```diff
@@ -1409,8 +1423,14 @@ module Shrinker = struct
       | Some c -> Iter.return c
       | _ -> Iter.empty)
     | ListTrm (t, lst, e) -> Iter.map (fun l -> ListTrm (t, l, e)) (Shrink.list lst)
-    | Constructor _ as c -> Iter.return c
-    | PatternMatch _ -> Iter.return term (* FIXME: not implemented *)
+    | Constructor (typ, name, args) ->
+      let open Iter in
+      list_elems term_shrinker args >|= fun args' -> Constructor (typ, name, args')
+    | PatternMatch (typ, matched_trm, cases, eff) ->
+      let open Iter in
+      term_shrinker matched_trm >>= fun matched_trm' ->
+      list_elems (fun (pat, body) -> Iter.pair (return pat) (term_shrinker body)) cases
+      >|= fun cases' -> PatternMatch (typ, matched_trm', cases', eff)
```

Comment:
This attempts to shrink all possible sub-parts, one after another which is a natural idea.
However, if one sub-part cannot be shrunk it will fail to shrink the others!
For this reason, it is general folklore to shrink each sub-part separately (ordering them from 'most aggressive' to 'least aggressive') and gluing them all together with `<+>` (sequentialize iterators)

Reply:
Thanks! Will fix.

How do you measure aggressiveness of shrinking though
?

Reply:
One way is to count/estimate shrinking steps for example inputs, but I suppose it also comes from experience. For example,
- an effective list shrinker should try cutting down the list size first, before proceeding with shrinking the individual elements (QCheck's list shrinker does this). Otherwise, you might waste time shrinking each element of list of 1000s of elements only to later remove them entirely
- to shrink an integer, one can attempt to halve it or to subtract 1 (for positive integers at least). Starting from `max_int` it is much fruitful to start with the halving strategy, and once that fails proceed with simple decrements

Reply:
You can achieve this "one then the other" behavior by using the `Shrink.pair : 'a Shrink.t -> 'b Shrink.t -> ('a * 'b) Shrink.t` function.

```ocaml
Shrink.pair
  term_shrinker (list_elem case_shrinker)
  (matched_trm, cases)
>|= fun (matched_trm', cases') ->
  ...
```